source/Calamari.Tests/Terraform/TerraformFixture.cs

```diff
@@ -0,0 +1,358 @@
+                }, "WithVariables")
+                .Should().Contain(actual);
+        }
+
+        [Test]
+        public void WithOutputSensitiveVariables()
+        {
+            ExecuteAndReturnLogOutput<ApplyCommand>(_ => { }, "WithOutputSensitiveVariables")
+                .Should().Contain("sensitive=\"");
+        }
+
+        [Test]
+        public void OutputAndSubstituteOctopusVariables()
+        {
+            ExecuteAndReturnLogOutput<ApplyCommand>(_ =>
+                {
+                    _.Set(TerraformSpecialVariables.Action.Terraform.VarFiles, "example.txt");
+                    _.Set(TerraformSpecialVariables.Action.Terraform.FileSubstitution, "example.txt");
+                    _.Set("Octopus.Action.StepName", "Step Name");
+                    _.Set("Should_Be_Substituted", "Hello World");
+                    _.Set("Should_Be_Substituted_in_txt", "Hello World from text");
+                }, "WithVariablesSubstitution")
+                .Should()
+                .Contain("Octopus.Action[\"Step Name\"].Output.TerraformValueOutputs[\"my_output\"]' with the value only of 'Hello World'")
+                .And
+                .Contain("Octopus.Action[\"Step Name\"].Output.TerraformValueOutputs[\"my_output_from_txt_file\"]' with the value only of 'Hello World from text'");
+        }
+
+        [Test]
+        [TestCase(typeof(PlanCommand))]
+        [TestCase(typeof(DestroyPlanCommand))]
+        public void TerraformPlanOutput(Type commandType)
+        {
+            ExecuteAndReturnLogOutput(commandType, _ =>
+                {
+                    _.Set("Octopus.Action.StepName", "Step Name");
+                }, "Simple")
+                .Should().Contain("Octopus.Action[\"Step Name\"].Output.TerraformPlanOutput");
+        }
+
+        [Test]
+        public void UsesWorkSpace()
+        {
+            ExecuteAndReturnLogOutput<ApplyCommand>(_ =>
+                {
+                    _.Set(TerraformSpecialVariables.Action.Terraform.Workspace, "myspace");
+                }, "Simple")
+                .Should().Contain("workspace new \"myspace\"");
+        }
+
+        [Test]
+        public void UsesTemplateDirectory()
+        {
+            ExecuteAndReturnLogOutput<ApplyCommand>(_ =>
+                {
+                    _.Set(TerraformSpecialVariables.Action.Terraform.TemplateDirectory, "SubFolder");
+                }, "TemplateDirectory")
+                .Should().Contain("SubFolder\\example.tf");
+        }
+
+        [Test]
+        public async Task AzureIntegration()
+        {
+            var random = Guid.NewGuid().ToString("N").Substring(0, 6);
+            var appName = $"cfe2e-{random}";
+
+            void PopulateVariables(VariableDictionary _)
+            {
+                _.Set(SpecialVariables.Action.Azure.SubscriptionId, Environment.GetEnvironmentVariable("Azure_OctopusAPITester_SubscriptionId"));
```

Comment:
It would be good if this test failed with a helpful message when certain environment variables aren't present. It should tell the developer where to get the secrets in LastPass.

Reply:
fixed